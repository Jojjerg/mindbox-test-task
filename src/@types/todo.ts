export interface ITodo {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export type TodoContextType = {
  todos: ITodo[];
  filter: IFilter;
  getTodos: () => void;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

export type IFilter = "All" | "Completed" | "Active";
