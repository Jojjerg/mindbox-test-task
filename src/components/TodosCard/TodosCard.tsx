import type { IFilter, ITodo } from "../../@types/todo";

import AddTodo from "../AddTodo/AddTodo";
import React from "react";
import TodosFilter from "../TodosFilter/TodosFilter";
import TodosList from "../TodosList/TodosList";

type CardProps = {
  todos: ITodo[];
  filter: IFilter;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

const TodosCard: React.FC<CardProps> = ({
  todos,
  filter,
  addTodo,
  updateTodo,
  deleteCompletedTodos,
  changeFilter,
}) => {
  return (
    <div data-testid="todos-card" className="flex flex-col gap-4 w-[1000px] h-max rounded-md shadow-lg bg-[#282828]">
      <AddTodo addTodo={addTodo} />
      <TodosList todos={todos} filter={filter} updateTodo={updateTodo} />
      <TodosFilter todos={todos} changeFilter={changeFilter} deleteCompletedTodos={deleteCompletedTodos} />
    </div>
  );
};

export default TodosCard;
