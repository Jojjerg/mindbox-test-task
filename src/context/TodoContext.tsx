import type { ITodo, TodoContextType } from "../@types/todo";
import { createContext, useContext, useState } from "react";

import type { PropsWithChildren } from "react";
import React from "react";

export const TodoContext = createContext<TodoContextType | null>(null);

const url = "https://jsonplaceholder.typicode.com/todos/?_limit=8";

const TodoProvider = ({ children }: PropsWithChildren<{}>) => {
  const [todos, setTodos] = useState<TodoContextType['todos']>([]);
  const [filter, setFilter] = useState<TodoContextType['filter']>("All");

  const getTodos = () => {
    try {
      fetch(url)
        .then((response) => response.json())
        .then((todos) => setTodos(todos));
    } catch (error) {
      console.error(error);
    }
  };

  const addTodo = (title: string) => {
    const newTodo: ITodo = {
      userId: Math.random(),
      id: Math.random(),
      title: title,
      completed: false,
    };

    setTodos([...todos, newTodo]);
  };

  const updateTodo = (id: number) => {
    todos.filter((todo: ITodo) => {
      if (todo.id === id) {
        if (todo.completed) {
          todo.completed = false;
        } else {
          todo.completed = true;
        }
        setTodos([...todos]);
      }
    });
  };

  const deleteCompletedTodos = () => {
    setTodos(todos.filter((todo: ITodo) => !todo.completed));
  };

  const changeFilter = (filter: TodoContextType['filter']) => {
    setFilter(filter);
  };

  const contextValue = {
    todos,
    filter,
    getTodos,
    addTodo,
    updateTodo,
    deleteCompletedTodos,
    changeFilter,
  };

  return (
    <TodoContext.Provider value={contextValue}>{children}</TodoContext.Provider>
  );
};

export const useTodoContext = () => {
  const context = useContext(TodoContext)
  
  if(!context) throw new Error("useTodoContext must be used inside the TodoProvider")

  return context
}

export default TodoProvider;
