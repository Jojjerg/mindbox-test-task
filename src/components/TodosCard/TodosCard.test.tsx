import { ITodo, TodoContextType } from "../../@types/todo";
import { render, screen } from "@testing-library/react";

import React from "react";
import TodosCard from "./TodosCard";
import { createRenderer } from "react-test-renderer/shallow";
import { useTodoContext } from "../../context/TodoContext";

const renderer = createRenderer();

const todos = [
  {
    userId: 1,
    id: 1,
    title: "delectus aut autem",
    completed: false,
  },
  {
    userId: 1,
    id: 2,
    title: "quis ut nam facilis et officia qui",
    completed: false,
  },
  {
    userId: 1,
    id: 3,
    title: "fugiat veniam minus",
    completed: false,
  },
  {
    userId: 1,
    id: 4,
    title: "et porro tempora",
    completed: true,
  },
  {
    userId: 1,
    id: 5,
    title: "laboriosam mollitia et enim quasi adipisci quia provident illum",
    completed: false,
  },
  {
    userId: 1,
    id: 6,
    title: "qui ullam ratione quibusdam voluptatem quia omnis",
    completed: false,
  },
  {
    userId: 1,
    id: 7,
    title: "illo expedita consequatur quia in",
    completed: false,
  },
  {
    userId: 1,
    id: 8,
    title: "quo adipisci enim quam ut ab",
    completed: true,
  },
];

let filter = "All";

const addTodo = (title: string) => {
  const newTodo: ITodo = {
    userId: Math.random(),
    id: Math.random(),
    title: title,
    completed: false,
  };

  todos.push(newTodo);

  return todos;
};

const updateTodo = (id: number) => {
  todos.filter((todo: ITodo) => {
    if (todo.id === id) {
      if (todo.completed) {
        todo.completed = false;
      } else {
        todo.completed = true;
      }
      return [...todos];
    }
  });
};

const deleteCompletedTodos = () => {
  return todos.filter((todo: ITodo) => !todo.completed);
};

const changeFilter = (newFilter: TodoContextType["filter"]) => {
  filter = newFilter;
  return filter;
};

const defaultComponent = (
  <TodosCard
    todos={todos}
    filter={"All"}
    addTodo={addTodo}
    updateTodo={updateTodo}
    changeFilter={changeFilter}
    deleteCompletedTodos={deleteCompletedTodos}
  />
);

describe("<TodosCard />", () => {
  it("should render and match the snapshot", () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
