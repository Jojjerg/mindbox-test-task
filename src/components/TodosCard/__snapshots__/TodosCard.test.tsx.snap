// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodosCard /> should render and match the snapshot 1`] = `
<div
  className="flex flex-col gap-4 w-[1000px] h-max rounded-md shadow-lg bg-[#282828]"
  data-testid="todos-card"
>
  <AddTodo />
  <TodosList />
  <TodosFilter />
</div>
`;
